src/java/org/apache/cassandra/schema/SchemaKeyspace.java

```diff
@@ -857,6 +891,36 @@ static Keyspaces fetchNonSystemKeyspaces()
         return fetchKeyspacesWithout(SchemaConstants.LOCAL_SYSTEM_KEYSPACE_NAMES);
     }
 
+    public static void validateNonCompact() throws StartupException
+    {
+        String query = format("SELECT keyspace_name, table_name, flags FROM %s.%s", SchemaConstants.SCHEMA_KEYSPACE_NAME, TABLES);
+
+        String messages = "";
+        for (UntypedResultSet.Row row : query(query))
+        {
+            if (SchemaConstants.isLocalSystemKeyspace(row.getString("keyspace_name")))
+                continue;
+
+            Set<String> flags = row.getSet("flags", AsciiType.instance);
+            if (TableMetadata.Flag.isLegacyCompactTable(TableMetadata.Flag.fromStringSet(flags)))
+            {
+                messages += String.format("ALTER TABLE %s.%s DROP COMPACT STORAGE;\n",
+                                          maybeQuote(row.getString("keyspace_name")),
+                                          maybeQuote(row.getString("table_name")));
+            }
+        }
+
+        if (!messages.isEmpty())
+        {
+            throw new StartupException(StartupException.ERR_OUTDATED_SCHEMA,
+                                       String.format("Compact Tables are not allowed in Cassandra starting with 4.0 version. " +
+                                                     "In order to migrate off Compact Storage, downgrade to the latest Cassandra version, " +
+                                                     "and run the following commands: \n\n%s\n" +
```

Comment:
Nice error message. Perhaps we could be even more clear specifying that the commands to run are CQL commands, given that this message is not shown in CQL:
```suggestion
                                                     "and run the following CQL commands: \n\n%s\n" +
```